src/Numerical/BLAS/Single.hs

```diff
@@ -249,3 +251,64 @@ isamax !n sx !incx
                       EQ -> findmax k c (i+1) (ix+incx)
                       LT -> f $ negate xi
                       GT -> f xi
+
+
+{- | O(n) sdot computes the sum of the products of elements drawn from two
+ vectors according to the following specification:
+
+@
+ sdsdot n u incx v incy = sum { u[f incx k] * v[f incy k] | k<=[0..n-1] }
+ where
+ f inc k | inc > 0  = inc * k
+         | inc < 0  = (1-n+k)*inc
+@
+The elements selected from the two vectors are controlled by the parameters
+n, incx and incy.   The parameter n determines the number of summands, while
+the parameters incx and incy determine the spacing between selected elements
+and the direction which the vectors are traversed.  When both incx and incy
+are unity and n is the length of both vectors then
+sdsdot corresponds to the dot product of two vectors.
+
+NOTE: The summation is performed using double precision floating point and the
+result is coerced to a single immediatly prior to returing the value.
+-}
+sdsdot :: Int -> Float -> Vector Float -> Int -> Vector Float -> Int -> Float
+sdsdot n (floatToDouble -> sb) sx !incx sy !incy
+   | n< 1                   = doubleToFloat sb
+   | incx==incy && incx > 0 = whenPositiveEqualInc (n*incx)
+   | otherwise              = whenUnequalOrNegIncs 0 sb (firstIndex n incx) (firstIndex n incy)
+   where
+   whenPositiveEqualInc :: Int -> Float
+   {-# INLINE whenPositiveEqualInc #-}
+   whenPositiveEqualInc !imax = whenPositiveEqualInc_loop 0 sb
+       where
+       whenPositiveEqualInc_loop :: Int -> Double -> Float
+       {-# INLINE whenPositiveEqualInc_loop #-}
+       whenPositiveEqualInc_loop !i !c
+            | i>=imax   = doubleToFloat c
+            | otherwise =
+                let c' = c + (floatToDouble ( sx `V.unsafeIndex` i))
+                           * (floatToDouble ( sy `V.unsafeIndex` i))
+                in whenPositiveEqualInc_loop (i+incx) c'
+   whenUnequalOrNegIncs :: Int -> Double -> Int -> Int -> Float
+   {-# INLINE whenUnequalOrNegIncs #-}
+   whenUnequalOrNegIncs !i !c !kx !ky
+       | i >= n    = doubleToFloat c
+       | otherwise =
+           let c' = c + (floatToDouble $ sx `V.unsafeIndex` kx)
+                      * (floatToDouble $ sy `V.unsafeIndex` ky)
+           in  whenUnequalOrNegIncs (i+1) c' (kx+incx) (ky+incy)
+
+
+--- UTILITIES --
+-- Move these to an appropriate module
```

Comment:
Below this line are the utility functions that were added to enable coercion between different IEEE floating point types.   If these are of general use, they should be moved to another module, otherwise they should be inlined.    For now we leave them here.